torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -0,0 +1,152 @@
+
+      // TODO: make this asynchronous
+      // src is only alive within this block, use reference to avoid copy
+      auto& stack = src->stackRef();
+      src->op()->getOperation()(stack);
+      AT_ASSERT(
+          stack.size() == 1,
+          "Return value of a builtin operator or a "
+          "TorchScript function should be a single IValue, got a vector of "
+          "size ",
+          stack.size());
+
+      ownerRRef->setValue(std::move(stack.front()));
+      return nullptr;
+    }
+    case MessageType::RREF_FETCH_CALL: {
+      auto srf = static_cast<ScriptRRefFetchCall*>(rpc);
+      // TODO: make this asynchronous
+      std::shared_ptr<OwnerRRef<IValue>> rref =
+          RRefContext::getInstance()->getOrCreateOwnerRRef<IValue>(
+              RRefId::fromIValue(srf->value()));
+      return std::unique_ptr<ScriptRRefFetchRet>(
+          new ScriptRRefFetchRet(rref->getValue()));
+    }
+    case MessageType::RREF_USER_CREATE: {
+      auto sra = static_cast<ScriptRRefCreate*>(rpc);
+      RRefContext::getInstance()->addFork(sra->valueRef());
+      return nullptr;
+    }
+    case MessageType::RREF_USER_DELETE: {
+      auto srd = static_cast<ScriptRRefDelete*>(rpc);
+      RRefContext::getInstance()->delFork(srd->valueRef());
+      return nullptr;
+    }
+    case MessageType::MESSAGE_WITH_AUTOGRAD_REQ: {
+      auto rpcWithAutograd = static_cast<RpcWithAutograd*>(rpc);
+      const auto& autogradMetadata = rpcWithAutograd->autogradMetadata();
+
+      // Attach 'recv' autograd function.
+      DistAutogradContext* autogradContext = addRecvRpcBackward(
+          rpcWithAutograd->autogradMetadata(), rpcWithAutograd->tensors());
+
+      // Process the original RPC.
+      auto wrappedRpcResponse = processRpc(
+          rpcWithAutograd->moveWrappedRpc().get(),
+          rpcWithAutograd->wrappedMessageType());
+
+      // Wrap the response with autograd, need a new autograd message id for
+      // each send/recv pair.
+      auto& autogradContainer = DistAutogradContainer::getInstance();
+      AutogradMetadata responseAutogradMetadata(
+          autogradMetadata.autogradContextId,
+          autogradContainer.newAutogradMessageId());
+
+      std::unique_ptr<RpcWithAutograd> response(new RpcWithAutograd(
+          MessageType::MESSAGE_WITH_AUTOGRAD_RESP,
+          responseAutogradMetadata,
+          std::move(wrappedRpcResponse)));
+
+      // Attach the 'send' autograd function if needed.
+      if (autogradContext != nullptr) {
+        addSendRpcBackward(
+            *autogradContext, responseAutogradMetadata, response->tensors());
+      }
+      return response;
+    }
+    default: {
+      TORCH_INTERNAL_ASSERT(
+          false, "Request type ", messageType, " not supported.");
+    }
+  }
+}
+
+Message RequestCallbackImpl::processMessage(Message& request) {
+  std::unique_ptr<RpcBase> rpc = deserializeRequest(request);
+  auto response = processRpc(rpc.get(), request.type());
```

Comment:
The above two lines both will run switch-case on the message type. Any reason for separating deserialization from processRpc?

Reply:
I added a TODO in `processRpc` to handle this. For now, we do this since I need to call `processRpc` recursively after unwrapped rpc_with_autograd.